src/tls/socket-io.c

```diff
@@ -76,17 +104,41 @@ wait_for_io (struct timespec *start,
   uint64_t remaining;
   int r;
 
+  debug (SOCKET_IO, "wait_for_io(%d, %u, %ju):", fd, (unsigned) events, (uintmax_t) timeout_us);
+
   if (!get_remaining_timeout (start, &remaining, timeout_us))
     return 0;
 
+  debug (SOCKET_IO, "  -> waiting for %jd", (uintmax_t) remaining);
+
   do
     r = poll (&pfd, 1, (remaining + 999) / 1000);
   while (r == -1 && errno == ENOENT);
 
+  debug (SOCKET_IO, "  -> result is %d/%s", r, r < 0 ? strerror (errno) : "-");
+
   return r;
 }
 
-bool
+/**
+ * recv_all:
+ * @fd: a file descriptor for a connected stream socket
+ * @buffer: a buffer
+ * @size: the size of @buffer
+ * @timeout: a timeout, in microseconds
+ *
+ * Attempts to read up to @size - 1 bytes from the connected stream
+ * socket @fd, followed by EOF.  On success, a nul terminator is
+ * inserted after the last byte and the number of bytes read (which
```

Comment:
It would make more sense if recv_all wouldn't do that, and recv_alnum() adds the nul-terminator instead.

Reply:
I tried to implement it this way, but quickly remembered my reason for not doing this: after we recv() the normal data, we need to recv() once more to ensure that we get a 0 (ie: EOF).  We need at least one extra byte to do that into, and the byte where we would insert the nul terminator is extremely convenient for that purpose.

Otherwise I'd need to have a separate `char b` variable or something and a separate `recv()` that tries to read into that to make sure we get EOF.

This is definitely "clever".  I could add a comment about it.